File: dijkstra.py
from math import inf

#declaration 
#graph is considered as unvisited
visited = [] # 
shortest_distance = {} # take the shortest node to reach destination
tracking = {} #searching for the optimal route
path_tracking = [] # path to the route (prev_node)

def assignInfinity(graph,start,end):
    #assign all nodes of shortest distance to inf except starting node
    for node in graph:
        shortest_distance[node] = inf
    shortest_distance[start] = 0


def process(graph,start,end):
    while graph:
        min_distance = None
        #check the index for the smallest distance of node 
        for node in graph:
            if min_distance is None:
                min_distance = node
            elif shortest_distance[node] < shortest_distance[min_distance]:
                min_distance = node
    
        path_options = graph[min_distance].items()

        tracking[start] = 0

        for child_node, weight in path_options:
            calc_distance = weight + shortest_distance[min_distance]
            if calc_distance < shortest_distance[child_node]:
                shortest_distance[child_node] = calc_distance
                tracking[child_node]= min_distance

        graph.pop(min_distance)
        visited.append(min_distance) 

def outputshortestdistance(graph,start,end):
    curNode = end
    while curNode != start:
        try:
            path_tracking.insert(0, curNode)
            curNode = tracking[curNode]
        except:
            print("Path does not exist")
            break
        
    path_tracking.insert(0, start)
    
    if shortest_distance[end] != inf:
        print("\nShortest distance is " + str(shortest_distance[end]))
        print("Path = " + str(path_tracking))
        print("\n------------------------------------Table------------------------------------------")
        print("Vertex\t|\tShortest distance from vertex " + start.upper() + "\t|\t Previous vertex")
        for node in tracking:
            print(str(node) + "\t\t" + str(shortest_distance[node])+ "\t\t\t\t\t\t" + str(tracking[node]))
    
    

File: Dijkstra.cpp
#include <iostream>
#define INF 999

using namespace std;

void output(int start, int dist[], int pred[])
{
    char outputnode[] ={'A','B','C','D','E','S'};
    for ( int x=0; x< 6; x++)
    {
        if ( x != start)
        {
            if( dist[x] != 999)
            {
                cout << "The distance of node " << outputnode[x] << " is " << dist[x] << endl;
                cout << outputnode[x];
                int j = x;
                do
                {
                    j = pred[j];
                    cout << " <- " << outputnode[j];
                }while(j != start);
                cout << endl;
            }
        }
    }
}

void dijkstra(int graph[6][6], int v, int start)
{
    int dist[INF], visited[INF], pred[INF], count, min_distance, nextnode = 0;
    
    for ( int x=0; x< 6; x++)
    {
        dist[x] = graph[start][x];
        pred[x] = start;
        visited[x] = 0;
    }
    
    dist[start] = 0;
    visited[start] = 1;
    count = 1;
    
    while(count < 6-1)
    {
        min_distance = INF;
        for ( int x=0; x< 6; x++)
        {
            if ( dist[x] < min_distance && !visited[x])
            {
                min_distance = dist[x];
                nextnode = x;
            }
        }
        
        visited[nextnode] = 1;
        
        for ( int x=0; x < 6; x++)
        {
            if ( !visited[x])
            {
                if ( (min_distance + graph[nextnode][x]) < dist[x])
                {
                    dist[x] =  min_distance + graph[nextnode][x];
                    pred[x] = nextnode;
                }
            }
        }
        dist[start] = 0;
        
        count++;
        
        
    }
    output(start, dist, pred);
}



int main()
{
    int src;
    int graph[6][6] = {
    {0,6,INF,INF,INF,INF},
    {INF,0,INF,1,2,INF},
    {4,INF,0,3,INF,INF},
    {INF,INF,INF,0,1,INF},
    {INF,INF,INF,INF,0,INF},
    {1,INF,2,INF,INF,0}
    };
    
    cout << "Enter source node: ";
    cin >> src;
    
    dijkstra(graph,6, src);
}


